File: package/rust/src/kmeans_clustering.rs
use std::f64;

pub fn kmeans_clustering(points: &[(f64, f64)], k: usize, max_iter: usize) -> Vec<usize> {
    if points.is_empty() || k == 0 {
        return vec![];
    }

    if k >= points.len() {
        return (0..points.len()).collect();
    }

    let mut centroids = initialize_centroids(points, k);
    let mut assignments = vec![0; points.len()];

    for _ in 0..max_iter {
        let new_assignments = assign_points_to_centroids(points, &centroids);

        if new_assignments == assignments {
            break;
        }

        assignments = new_assignments;
        centroids = update_centroids(points, &assignments, k);
    }

    assignments
}

fn initialize_centroids(points: &[(f64, f64)], k: usize) -> Vec<(f64, f64)> {
    let mut centroids = Vec::new();
    let step = points.len() / k;

    for i in 0..k {
        let index = i * step;
        centroids.push(points[index]);
    }

    centroids
}

fn assign_points_to_centroids(points: &[(f64, f64)], centroids: &[(f64, f64)]) -> Vec<usize> {
    points.iter().map(|point| {
        centroids.iter()
            .enumerate()
            .min_by(|(_, c1), (_, c2)| {
                let d1 = distance(point, c1);
                let d2 = distance(point, c2);
                d1.partial_cmp(&d2).unwrap()
            })
            .map(|(i, _)| i)
            .unwrap_or(0)
    }).collect()
}

fn update_centroids(points: &[(f64, f64)], assignments: &[usize], k: usize) -> Vec<(f64, f64)> {
    (0..k).map(|cluster| {
        let cluster_points: Vec<_> = points.iter()
            .zip(assignments.iter())
            .filter(|(_, &assignment)| assignment == cluster)
            .map(|(point, _)| point)
            .collect();

        if cluster_points.is_empty() {
            (0.0, 0.0)
        } else {
            let sum_x: f64 = cluster_points.iter().map(|(x, _)| x).sum();
            let sum_y: f64 = cluster_points.iter().map(|(_, y)| y).sum();
            let count = cluster_points.len() as f64;
            (sum_x / count, sum_y / count)
        }
    }).collect()
}

fn distance(p1: &(f64, f64), p2: &(f64, f64)) -> f64 {
    let dx = p1.0 - p2.0;
    let dy = p1.1 - p2.1;
    (dx * dx + dy * dy).sqrt()
}

File: package/php/src/KMeansClustering.php
<?php

declare(strict_types=1);

namespace ProgramKnock;

class KMeansClustering
{
    /**
     * K-meansクラスタリングを実行
     *
     * @param array<array{0: float, 1: float}> $points データ点のリスト
     * @param int $k クラスタ数
     * @param int $maxIter 最大反復回数
     * @return array<int> 各データ点が属するクラスタのインデックス
     */
    public static function cluster(array $points, int $k, int $maxIter = 100): array
    {
        if (empty($points)) {
            return [];
        }

        if ($k === 1) {
            return array_fill(0, count($points), 0);
        }

        $n = count($points);
        $k = min($k, $n);

        // 重心を初期化（最初のk個の点を使用）
        $centroids = [];
        for ($i = 0; $i < $k; $i++) {
            $centroids[] = [$points[$i][0], $points[$i][1]];
        }

        $assignments = array_fill(0, $n, 0);

        for ($iter = 0; $iter < $maxIter; $iter++) {
            $newAssignments = [];

            // 各点を最も近い重心に割り当て
            for ($i = 0; $i < $n; $i++) {
                $minDistance = PHP_FLOAT_MAX;
                $closestCluster = 0;

                for ($j = 0; $j < $k; $j++) {
                    $distance = self::euclideanDistance($points[$i], $centroids[$j]);
                    if ($distance < $minDistance) {
                        $minDistance = $distance;
                        $closestCluster = $j;
                    }
                }

                $newAssignments[$i] = $closestCluster;
            }

            // 収束チェック
            if ($newAssignments === $assignments) {
                break;
            }

            $assignments = $newAssignments;

            // 重心を再計算
            $newCentroids = [];
            for ($j = 0; $j < $k; $j++) {
                $clusterPoints = [];
                for ($i = 0; $i < $n; $i++) {
                    if ($assignments[$i] === $j) {
                        $clusterPoints[] = $points[$i];
                    }
                }

                if (!empty($clusterPoints)) {
                    $sumX = 0.0;
                    $sumY = 0.0;
                    foreach ($clusterPoints as $point) {
                        $sumX += $point[0];
                        $sumY += $point[1];
                    }
                    $count = count($clusterPoints);
                    $newCentroids[$j] = [$sumX / $count, $sumY / $count];
                } else {
                    // 空のクラスタの場合、元の重心を維持
                    $newCentroids[$j] = $centroids[$j];
                }
            }

            $centroids = $newCentroids;
        }

        return $assignments;
    }

    /**
     * ユークリッド距離を計算
     *
     * @param array{0: float, 1: float} $point1 点1
     * @param array{0: float, 1: float} $point2 点2
     * @return float ユークリッド距離
     */
    private static function euclideanDistance(array $point1, array $point2): float
    {
        $dx = $point1[0] - $point2[0];
        $dy = $point1[1] - $point2[1];
        return sqrt($dx * $dx + $dy * $dy);
    }
}
